consensus/src/memory_pool.rs

```diff
@@ -18,244 +18,158 @@
-                .fetch_sub(entry.size_in_bytes, Ordering::SeqCst);
+    //     Ok(())
+    // }
 
-            let transaction_id = entry.transaction.transaction_id()?.to_vec();
+    // /// Removes transaction from memory pool or error.
+    // #[inline]
+    // pub async fn remove(&self, entry: &Entry) -> Result<Option<Vec<u8>>, ConsensusError> {
+    //     if self.contains(entry) {
+    //         let transaction_id = entry.transaction.transaction_id()?.to_vec();
 
-            self.transactions.remove(transaction_id.to_vec()).await;
+    //         self.transactions.remove(transaction_id.to_vec()).await;
 
-            return Ok(Some(transaction_id));
-        }
+    //         return Ok(Some(transaction_id));
+    //     }
 
-        Ok(None)
-    }
+    //     Ok(None)
+    // }
 
     /// Removes transaction from memory pool based on the transaction id.
-    #[inline]
-    pub async fn remove_by_hash(&self, transaction_id: &[u8]) -> Result<Option<Entry<T>>, ConsensusError> {
-        match self.transactions.get(transaction_id) {
+    pub fn remove(&mut self, transaction_id: &Digest) -> Result<Option<SerialTransaction>, ConsensusError> {
+        match self.transactions.remove(transaction_id) {
             Some(entry) => {
-                self.total_size_in_bytes
-                    .fetch_sub(entry.size_in_bytes, Ordering::SeqCst);
-
-                self.transactions.remove(transaction_id.to_vec()).await;
-
-                Ok(Some(entry.clone()))
+                for SerialRecord { commitment, .. } in &entry.transaction.new_records {
+                    if !self.commitments.remove(commitment) {
+                        panic!("missing commitment from memory pool during removal");
+                    }
+                }
+                for serial in &entry.transaction.old_serial_numbers {
+                    if !self.serial_numbers.remove(serial) {
+                        panic!("missing serial from memory pool during removal");
+                    }
+                }
+                if !self.memos.remove(&entry.transaction.memorandum) {
+                    panic!("missing memo from memory pool during removal");
+                }
+                Ok(Some(entry.transaction))
             }
             None => Ok(None),
         }
     }
 
-    /// Returns whether or not the memory pool contains the entry.
-    #[inline]
-    pub fn contains(&self, entry: &Entry<T>) -> bool {
-        match &entry.transaction.transaction_id() {
-            Ok(transaction_id) => self.transactions.contains_key(&transaction_id.to_vec()),
-            Err(_) => false,
-        }
-    }
-
     /// Get candidate transactions for a new block.
-    pub fn get_candidates<P: LoadableMerkleParameters, S: Storage>(
-        &self,
-        storage: &Ledger<T, P, S>,
-        max_size: usize,
-    ) -> Result<DPCTransactions<T>, ConsensusError> {
+    pub fn get_candidates(&self, max_size: usize) -> Result<Vec<&SerialTransaction>, ConsensusError> {
```

Comment:
This method doesn't need to return a `Result` (and it doesn't look like it did before), and since its return value is always iterated over, it would be best if it returned an `impl Iterator`, if possible.

Reply:
i generally only do that for input types, for output types I leave as concrete unless there is an important reason not to (using a private type for example). main rationale is to preserve functionality and avoid in the future doing a vec -> into_iter -> collect(vec) chain by accident/etc.